pkg/networkservice/common/resourcepool/server.go

```
@@ -0,0 +1,191 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package resourcepool provides chain elements for to select and free VF
+package resourcepool
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/vfio"
+	"github.com/networkservicemesh/sdk-kernel/pkg/kernel/networkservice/vfconfig"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+
+	"github.com/networkservicemesh/sdk-sriov/pkg/sriov"
+)
+
+// PCIResourcePool is a resourcepool.ResourcePool + sync.Locker interface
+type PCIResourcePool interface {
+	Select(driverType sriov.DriverType, service string, capability sriov.Capability) (string, error)
+	Free(vfPciAddr string) error
+
+	sync.Locker
+}
+
+type resourcePoolServer struct {
+	driverType  sriov.DriverType
+	functions   map[sriov.PCIFunction][]sriov.PCIFunction
+	binders     map[uint][]sriov.DriverBinder
+	selectedVFs map[string]string
+}
+
+// NewServer returns a new resource pool server chain element
+func NewServer(driverType sriov.DriverType, functions map[sriov.PCIFunction][]sriov.PCIFunction, binders map[uint][]sriov.DriverBinder) networkservice.NetworkServiceServer {
+	return &resourcePoolServer{
+		driverType:  driverType,
+		functions:   functions,
+		binders:     binders,
+		selectedVFs: map[string]string{},
+	}
+}
+
+func (s *resourcePoolServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	logEntry := log.Entry(ctx).WithField("resourcePoolServer", "Request")
+
+	service, capability, err := getServiceAndCapability(request)
```

I think you need to take ResourcePool here and check it passed, fail fast in case if not.

---

Yep